src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,214 @@
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   */
+const nodeChange = (tree, oldPath, newPath) => {
+  const oldContext = pathToContext(oldPath)
+  const newContext = pathToContext(newPath)
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+
+  if (commonNode) {
+    if (commonNode.leaf) {
+      // A is changed to AF --> AF , A and A.B --> A.B , A.B and A.B.C --> A.B.C
+      _.unset(tree, commonPath)
+      _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+    }
+    else {
+      const leafNodes = findTreeDescendants(tree, commonPath)
+      // if oldPath is already available we just need to update its descendants
+      if (equalArrays(commonPath, oldContext)) {
+        // this flag is needed to know if ancestor replaces very old descentdant in the tree
+        let descendantReplaced = false // eslint-disable-line fp/no-let
+        // If A.E which already available in tree is updated to A.EM the updating just its descendants like A.E.O.M to A.EM.O.M.
+        leafNodes.forEach(descendant => {
+          if (timeDifference(timestamp(), descendant.lastUpdated) > EDIT_TIME_MAX) {
+            const descendantContext = pathToContext(descendant.path)
+            const closestAncestor = findClosestAncestorWithMultipleChildren(tree, descendantContext).path
+            _.unset(tree, descendantContext.slice(0, closestAncestor.length + 1))
+            descendantReplaced = true
+          }
+          else {
+            const updatedDescendantPath = newPath.concat(descendant.path.slice(newPath.length))
+            _.unset(tree, pathToContext(descendant.path))
+            _.set(tree, pathToContext(updatedDescendantPath), { leaf: true, lastUpdated: timestamp(), path: updatedDescendantPath })
+          }
+        })
+        if (descendantReplaced) nodeChange(tree, oldPath, newPath) // called once again to remove merge inconsitencty that might occur while replacing descendants by ancestor
+        else if (oldContext.slice(-1)[0] !== newContext.slice(-1)[0]) _.unset(tree, oldContext)
+      }
+      else {
+        let isMerged = false // eslint-disable-line fp/no-let
+        leafNodes.forEach(descendant => {
+          const descendantContext = pathToContext(descendant.path)
+          if (descendantContext[0] === '__EM__') return // preventing nodes at level 0 from merged to this (temporary fix)
+          const longContext = descendantContext.length > newContext.length ? descendantContext : newContext
+          const shortContext = newContext.length < descendantContext.length ? newContext : descendantContext
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (descendantContext.length === newContext.length && newContext.length - commonPath.length === 1) {
```

Comment:
You may want to consider creating constants for parts of conditionals to make them more readable. (Tell me if the names don't make sense here, I could be misunderstanding them, but you get the point.)

```js
const isSameDepth = descendantContext.length === newContext.length
const hasSameParent = newContext.length - commonPath.length === 1
const isSibling === isSameDepth && hasSameParent
...
```

This can be applied to other places to make the different conditions a bit clearer.

Reply:
It makes perfect sense. I will make these changes to other places.